traffic_ops/testing/api/v14/config/cache_config/cache_config.go

```diff
@@ -0,0 +1,265 @@
+/*
+   Licensed under the Apache License, Version 2.0 (the "License");
+   you may not use this file except in compliance with the License.
+   You may obtain a copy of the License at
+
+   http://www.apache.org/licenses/LICENSE-2.0
+
+   Unless required by applicable law or agreed to in writing, software
+   distributed under the License is distributed on an "AS IS" BASIS,
+   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+   See the License for the specific language governing permissions and
+   limitations under the License.
+*/
+
+package cache_config
+
+import (
+	"regexp"
+	"strings"
+
+	. "github.com/apache/trafficcontrol/traffic_ops/testing/api/v14/config"
+	"github.com/apache/trafficcontrol/traffic_ops/traffic_ops_golang/test"
+	"github.com/go-ozzo/ozzo-validation/is"
+)
+
+// ParseCacheConfig takes a string presumed to be an ATS cache.config and validates
+// it is syntatically correct.
+//
+// The general format of a cache config is three types of labels separated by spaces:
+//
+//  primary_destination=value secondary_specifier=value action=value
+//
+// For a full description of how to format a cache config, refer to the ATS documentation
+// for the cache config:
+// https://docs.trafficserver.apache.org/en/latest/admin-guide/files/cache.config.en.html
+//
+func ParseCacheConfig(config string) test.Error {
+	lines := strings.Split(config, "\n")
+
+	if len(lines) == 1 {
+		return parseConfigRule(lines[0])
+	}
+
+	for i, ln := range lines {
+		err := parseConfigRule(ln)
+		if err != nil {
+			return err.Prepend("error on line %d: ", i+1)
+		}
+	}
+
+	return nil
+}
+
+func parsePrimaryDestinations(lhs string, rhs string) test.Error {
+
+	switch lhs {
+	case "dest_domain":
+		// dest_host is an alias for dest_domain
+		fallthrough
+	case "dest_host":
+		if err := is.Host.Validate(rhs); err != nil {
+			return ErrorContext.NewError(InvalidHost, `"%s" %v`, rhs, err)
+		}
+	case "dest_ip":
+		if err := is.IP.Validate(rhs); err != nil {
+			return ErrorContext.NewError(InvalidIP, `"%s" %v`, rhs, err)
+		}
+	case "host_regex":
+		fallthrough
+	case "url_regex":
+		// only makes sure the regex compiles, not that the regex generates anything valid
+		if _, err := regexp.Compile(rhs); err != nil {
+			return ErrorContext.NewError(InvalidRegex, "%v", err)
+		}
+	default:
+		return ErrorContext.NewError(InvalidLabel)
+	}
+
+	return nil
+}
+
+func parseSecondarySpecifiers(lhs string, rhs string) test.Error {
+
+	switch lhs {
+	case "port":
+		if err := is.Port.Validate(rhs); err != nil {
+			return ErrorContext.AddErrorCode(InvalidPort, err)
+		}
+	case "scheme":
+		if rhs != "http" && rhs != "https" {
+			return ErrorContext.NewError(InvalidHTTPScheme)
+		}
+	case "prefix":
+		// idk what validation to do on this
```

Comment:
This should have `TODO` to make it obvious to future readers that this should be solved, e.g. `// TODO determine what validation to do on this`

Reply:
To be clear: It is more of a comment to the reviewer than making a TODO that should be followed up on later. I don't think there is a good solution for validating the prefix label, unless a heuristic exists, and I don't think the test needs to be quite that granular. Although TODOs can be useful, I think they can often be forgotten about, and there is no mention of who is responsible for fulfilling a TODO statement.

Reply:
Ah. I'd say if you're sure there's no obvious validation to do, a comment stating that would be better, e.g. `case "prefix": // no logical validation`. If you're unsure, a TODO tells a future reader that adding something would be good, if they happen to know what would be good.

>they can often be forgotten about, and there is no mention of who is responsible for fulfilling a TODO statement.

Right, they shouldn't be used for important things, anything that would be a big problem if forgotten should be a Github issue. I like to use "TODO" for things that aren't super-important, but would be good if someone happens to notice, and has the time or knows the right answer.

Including the text "TODO" makes it possible to search for it. I often search for "TODO," to see if there's anything I know or can quickly fix.

But again, if there's no logical validation, IMO a comment to that effect would be good, so future readers know why the `case` is empty.